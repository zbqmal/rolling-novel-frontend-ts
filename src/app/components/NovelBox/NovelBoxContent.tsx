"use client";

import React, { useState } from "react";
import { NovelBoxContentState } from "./types";
import { NovelBoxContentStyles } from "./styles";
import { useRouter } from "next/navigation";

interface NovelBoxContentProps {
  id: string;
  title: string;
  description: string;
}

const NovelBoxContent: React.FC<NovelBoxContentProps> = ({
  id,
  title,
  description,
}) => {
  const router = useRouter();
  const [novelBoxContentState, setNovelBoxContentState] =
    useState<NovelBoxContentState>({
      title,
      description,
    });

  const pushToNovelDetailsPage = () => router.push(`novel/${id}`);

  return (
    <div
      style={NovelBoxContentStyles.box}
      onClick={() => pushToNovelDetailsPage()}
      data-testid={`novel-box-container-${id}`}
    >
      <h2 style={NovelBoxContentStyles.title}>{novelBoxContentState.title}</h2>
      <p style={NovelBoxContentStyles.description}>
        {novelBoxContentState.description}
      </p>
    </div>
  );
};

export default NovelBoxContent;
