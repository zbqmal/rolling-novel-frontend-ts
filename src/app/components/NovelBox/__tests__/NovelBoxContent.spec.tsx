import { render, screen, fireEvent } from "@testing-library/react";
import NovelBoxContent from "../NovelBoxContent";
import React from "react";
import { useRouter } from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

describe("NovelBoxContent", () => {
  const testId = "test-id";
  const testTitle = "test-title";
  const testDescription = "test-description";

  const mockPush = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <NovelBoxContent
        id={testId}
        title={testTitle}
        description={testDescription}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe("when novel box is clicked", () => {
    it("should navigate to its details page", () => {
      render(
        <NovelBoxContent
          id={testId}
          title={testTitle}
          description={testDescription}
        />
      );
      fireEvent.click(screen.getByTestId(`novel-box-container-${testId}`));

      expect(mockPush).toHaveBeenCalledWith(`novel/${testId}`);
    });
  });
});
