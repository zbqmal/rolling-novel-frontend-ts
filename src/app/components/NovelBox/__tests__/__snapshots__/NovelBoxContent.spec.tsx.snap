// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NovelBoxContent matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="novel-box-container-test-id"
    style="border: 1px solid #ccc; border-radius: 12px; padding: 16px; margin: 16px auto; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); height: 400px; width: 300px;"
  >
    <h2
      style="margin: 0px 0px 8px 0px; font-size: 18px; font-weight: bold; color: rgb(51, 51, 51);"
    >
      test-title
    </h2>
    <p
      style="margin: 0px; font-size: 14px; color: rgb(85, 85, 85);"
    >
      test-description
    </p>
  </div>
</DocumentFragment>
`;
